BacktraceScreen.c

```diff
@@ -0,0 +1,405 @@
+   }
+
+   free(objectTrimmed);
+}
+
+static void BacktraceFrame_display(const Object* super, RichString* out) {
+   const BacktraceFrame* const frame = (const BacktraceFrame* const)super;
+   assert(frame);
+
+   if (frame->index == BACKTRACE_FRAME_HEADER_INDEX) {
+      BacktraceFrame_displayHeader(frame, out);
+      return;
+   }
+
+   const BacktracePanelPrintingHelper* const printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   char* functionName = frame->functionName;
+   int maxFunctionNameLength = printingHelper->maxFunctionNameLength;
+   if ((displayOptions & DEMANGLE_NAME_FUNCTION) == DEMANGLE_NAME_FUNCTION &&
+         printingHelper->maxDemangledFunctionNameLength > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFunctionNameLength;
+      if (frame->demangleFunctionName) {
+         functionName = frame->demangleFunctionName;
+      }
+   }
+
+   xAsprintf(&functionName, "%s+0x%lx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjectNameLength;
+   if ((displayOptions & SHOW_FULL_PATH_OBJECT) == SHOW_FULL_PATH_OBJECT) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjectPathLength;
+   }
+
+   char* line = NULL;
+   int objectPathStart = -1;
+   int len = xAsprintf(&line, "%-*d 0x%0*lx %-*s %n%-*s",
+                       (int)printingHelper->maxNumberFrameLength, frame->index,
+                       (int)printingHelper->maxAddressLength - (int)strlen("0x"), frame->address,
+                       (int)maxFunctionNameLength, functionName,
+                       &objectPathStart,
+                       (int)objectLength, objectDisplayed ? objectDisplayed : "-");
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   if (functionName) {
+      free(functionName);
+   }
+
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* const frame1 = (const BacktraceFrame* const)object1;
+   const BacktraceFrame* const frame2 = (const BacktraceFrame* const)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
+}
+
+static void BacktraceFrame_delete(Object* object) {
+   BacktraceFrame* this = (BacktraceFrame*)object;
+   if (this->functionName) {
+      free(this->functionName);
+   }
+
+   if (this->demangleFunctionName) {
+      free(this->demangleFunctionName);
+   }
+
+   if (this->objectPath) {
+      free(this->objectPath);
+   }
+
+   if (this->objectName) {
+      free(this->objectName);
+   }
```

Comment:
You can assume `free` to be `NULL`-safe …

```suggestion
   free(this->functionName);
   free(this->demangleFunctionName);
   free(this->objectPath);
   free(this->objectName);
```